components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java

```diff
@@ -284,62 +278,122 @@ public WorkflowResponse updateApplication(String uuid, Application application)
+        OAuthAppRequest oauthAppRequest = new OAuthAppRequest(application.getName(), callbackUrl, tokenType,
                 grantTypes);
-        oauthAppRequest.getOAuthApplicationInfo().addParameter(KeyManagerConstants.VALIDITY_PERIOD, validityTime);
-        oauthAppRequest.getOAuthApplicationInfo().addParameter(KeyManagerConstants.APP_KEY_TYPE, tokenType);
-        oauthAppRequest.getOAuthApplicationInfo().addParameter(KeyManagerConstants.TOKEN_SCOPES, tokenScopes);
-        KeyManager keyManager = APIManagerFactory.getInstance().getKeyManager();
-
-        OAuthApplicationInfo oauthAppInfo = keyManager.createApplication(oauthAppRequest);
-        Map<String, Object> keyDetails = new HashMap<>();
-        if (oauthAppInfo != null) {
-            APIUtils.logDebug("Successfully created OAuth application", log);
-            keyDetails.put(KeyManagerConstants.KeyDetails.CONSUMER_KEY, oauthAppInfo.getClientId());
-            keyDetails.put(KeyManagerConstants.KeyDetails.CONSUMER_SECRET, oauthAppInfo.getClientSecret());
-            keyDetails.put(KeyManagerConstants.KeyDetails.SUPPORTED_GRANT_TYPES, oauthAppInfo.getGrantTypes());
-            keyDetails.put(KeyManagerConstants.KeyDetails.APP_DETAILS, oauthAppInfo.getJSONString());
-        } else {
-            throw new KeyManagementException("Error occurred while creating OAuth application");
+
+        OAuthApplicationInfo oauthAppInfo = getIdentityProvider().createApplication(oauthAppRequest);
+
+        if (log.isDebugEnabled()) {
+            log.debug("Application key generation was successful for application: " + application.getName()
+                    + " Client Id: " + oauthAppInfo.getClientId());
         }
 
-        AccessTokenRequest accessTokenRequest = ApplicationUtils.createAccessTokenRequest(oauthAppInfo);
-        AccessTokenInfo accessTokenInfo = keyManager.getNewAccessToken(accessTokenRequest);
-        // adding access token information to key details
-        if (accessTokenInfo != null) {
-            APIUtils.logDebug("Successfully created OAuth access token", log);
-            keyDetails.put(KeyManagerConstants.KeyDetails.ACCESS_TOKEN, accessTokenInfo.getAccessToken());
-            keyDetails.put(KeyManagerConstants.KeyDetails.VALIDITY_TIME, accessTokenInfo.getValidityPeriod());
-            keyDetails.put(KeyManagerConstants.KeyDetails.TOKEN_SCOPES, accessTokenInfo.getScopes());
-        } else {
-            throw new KeyManagementException("Error occurred while generating access token for OAuth application");
+        try {
+            getApplicationDAO().addApplicationKeys(applicationId, tokenType, oauthAppInfo);
+        } catch (APIMgtDAOException e) {
+            String errorMsg = "Error occurred while saving key data for application: " + application.getName();
+            log.error(errorMsg, e);
+            throw new APIManagementException(errorMsg, e, ExceptionCodes.APIMGT_DAO_EXCEPTION);
```

Comment:
shouldn't we just pass the exception code that is set from the DAO layer than setting it explicitly in the Service Impl layer? This should be corrected in all relevant places if so.